Shows top rationales for a given question

pomegranited opened the pull request:
* Adds peerinst.admin_views.get_question_rationale_aggregates() and unit tests
* Adds peerinst.admin_views.QuestionRationaleView to display rationale, using
  question_rationales.html template
* Adds link to the individual questions table in the assignment results admin page

**JIRA tickets**: [OC-1536](https://tasks.opencraft.com/browse/OC-1536)

**Screenshots**:
![screen shot 2016-04-26 at 6 06 22 pm](https://cloud.githubusercontent.com/assets/7556571/14811892/aa9b872e-0bd9-11e6-8e46-a2f79a8f2148.png)
![screen shot 2016-04-26 at 6 06 13 pm](https://cloud.githubusercontent.com/assets/7556571/14811897/b0d91566-0bd9-11e6-94db-7417e2819466.png)

**Testing instructions**:

1. Follow instructions in [README](https://github.com/open-craft/dalite-ng/blob/jill/fix-readme/README.md) to set up development environment and create a superuser.
1. [Run tests and verify coverage](https://github.com/open-craft/dalite-ng/blob/jill/fix-readme/README.md#running-the-tests)
1. Optionally load the test fixture data, or create your own assignments, questions, and responses.
```
./manage.py loaddata peerinst/fixtures/peerinst_test_data.yaml
``` 
1. Run the development server
```
./manage.py runserver
``` 
1. Navigate to http://127.0.0.1:8000/admin/ and click on an assignment under 'Student progress by assignment'.  From there, find the question table, and click on 'rationales' to see the new report.

**Reviewers**
- [ ] OpenCraft reviewer TBD

jbzdak commented on peerinst/admin_views.py:
```diff
@@ -86,6 +86,166 @@ def get_assignment_aggregates(assignment):
+            # rationales marked as chosen_rationales for answers that went from right to wrong,
+            # sorted by the number of times that rationale was chosen.
+            #
+            'right_to_wrong': [
```
This is the same section as starting from line [106](https://github.com/open-craft/dalite-ng/pull/43/files#diff-6fdfe2e57a137ae571d8a456bc39f4bdR106).

pomegranited replied:
Addressed by [1191cea L114](https://github.com/open-craft/dalite-ng/pull/43/commits/1191cea2af37032acdab2227eed74100b78e9b5d#diff-6fdfe2e57a137ae571d8a456bc39f4bdL114).

jbzdak commented on peerinst/admin_views.py:
```diff
@@ -86,6 +86,166 @@ def get_assignment_aggregates(assignment):
+    """Get aggregate upvoted rationales for the given assignment and question.
+
+    This function returns a pair (sums, rationales), where 'sums' is a collections.Counter object
+    mapping labels to integers, and 'rationales' is a dict containing a sorted list of dicts:
```
`mapping labels to integers` -> `mapping question labels to up vote counts`

pomegranited replied:
Addressed by [1191cea L93](https://github.com/open-craft/dalite-ng/pull/43/commits/1191cea2af37032acdab2227eed74100b78e9b5d#diff-6fdfe2e57a137ae571d8a456bc39f4bdL93).

jbzdak commented on peerinst/admin_views.py:
```diff
@@ -86,6 +86,166 @@ def get_assignment_aggregates(assignment):
+    This function returns a pair (sums, rationales), where 'sums' is a collections.Counter object
+    mapping labels to integers, and 'rationales' is a dict containing a sorted list of dicts:
+        {
+            # rationales where upvoted != rationales where upvoted !=0, sorted by upvoted
```
I don't understand this condition.

pomegranited replied:
That's because it's nonsensical.. sorry.  Fixed in [1191cea L95](https://github.com/open-craft/dalite-ng/pull/43/commits/1191cea2af37032acdab2227eed74100b78e9b5d#diff-6fdfe2e57a137ae571d8a456bc39f4bdL95)

jbzdak commented on peerinst/admin_views.py:
```diff
@@ -86,6 +86,166 @@ def get_assignment_aggregates(assignment):
+        }
+
+    """
+    answers = question.answer_set.filter(assignment=assignment).exclude(user_token='')
```
Why exclude empty `user_token=''? Please add an explanation if this is something non-intuitive.

pomegranited replied:
Fixed in [1191cea R124](https://github.com/open-craft/dalite-ng/pull/43/commits/1191cea2af37032acdab2227eed74100b78e9b5d#diff-6fdfe2e57a137ae571d8a456bc39f4bdR124)

jbzdak commented on peerinst/admin_views.py:
```diff
@@ -86,6 +86,166 @@ def get_assignment_aggregates(assignment):
+    answerchoice_correct = question.answerchoice_set.values_list('correct', flat=True)
+    correct_choices = list(itertools.compress(itertools.count(1), answerchoice_correct))
+
+    # Group by chosen_rational
```
chosen_rational -> chosen_rationale (add e)

pomegranited replied:
Fixed in [1191cea R142](https://github.com/open-craft/dalite-ng/pull/43/commits/1191cea2af37032acdab2227eed74100b78e9b5d#diff-6fdfe2e57a137ae571d8a456bc39f4bdR142)

jbzdak commented on peerinst/admin_views.py:
```diff
@@ -86,6 +86,166 @@ def get_assignment_aggregates(assignment):
     return sums, question_data
 
 
+def get_question_rationale_aggregates(assignment, question):
```
Maybe add 'controller' file, as this is not a view.

pomegranited replied:
Yep, I agree, but I'm sticking with the current convention.

jbzdak commented on peerinst/admin_views.py:
```diff
@@ -86,6 +86,166 @@ def get_assignment_aggregates(assignment):
+
+    # Group by chosen_rational
+    rationales = {a.id: a for a in answers.all()}
+    rationales[None] = None
```
Add a comment why `None` is needed here.

pomegranited replied:
Refactored out of [1191cea](https://github.com/open-craft/dalite-ng/pull/43/commits/1191cea2af37032acdab2227eed74100b78e9b5d)

jbzdak commented on peerinst/admin_views.py:
```diff
@@ -86,6 +86,166 @@ def get_assignment_aggregates(assignment):
+    }
+
+    # Calculate the sums: total number of rationales for each group
+    sums = collections.Counter(
```
Why not plain map, this looks to me as a normal plain map --- no counter specific operations are used here --- like addition of whole counter.

pomegranited replied:
Refactored out of [1191cea](https://github.com/open-craft/dalite-ng/pull/43/commits/1191cea2af37032acdab2227eed74100b78e9b5d)

jbzdak commented:
It is possible to generate 500 HTTP error by: 

1. Add a question
2. Add a predefined rationale `Admin` -> `Add Question` -> `Save and continue editing` -> `Top right corner: Previev` -> `Add add rationales`. 
3. Fill the questionaire, select an answer, and then change it using quoting a predefined rationale`

```
Environment:


Request Method: GET
Request URL: http://localhost:10100/admin/peerinst/assignment_results/test%3Dreport/rationales/7

Django Version: 1.8.12
Python Version: 2.7.11
Installed Applications:
('grappelli',
 'django.contrib.admin',
 'django.contrib.auth',
 'django.contrib.contenttypes',
 'django.contrib.sessions',
 'django.contrib.messages',
 'django.contrib.staticfiles',
 'peerinst',
 'django_lti_tool_provider')
Installed Middleware:
('django.contrib.sessions.middleware.SessionMiddleware',
 'django.middleware.common.CommonMiddleware',
 'django.middleware.csrf.CsrfViewMiddleware',
 'django.contrib.auth.middleware.AuthenticationMiddleware',
 'django.contrib.auth.middleware.SessionAuthenticationMiddleware',
 'django.contrib.messages.middleware.MessageMiddleware',
 'django.middleware.security.SecurityMiddleware')


Traceback:
File "/home/jboc/programs/dalite-ng/venv/lib/python2.7/site-packages/django/core/handlers/base.py" in get_response
  132.                     response = wrapped_callback(request, *callback_args, **callback_kwargs)
File "/home/jboc/programs/dalite-ng/venv/lib/python2.7/site-packages/django/contrib/auth/decorators.py" in _wrapped_view
  22.                 return view_func(request, *args, **kwargs)
File "/home/jboc/programs/dalite-ng/venv/lib/python2.7/site-packages/django/views/generic/base.py" in view
  71.             return self.dispatch(request, *args, **kwargs)
File "/home/jboc/programs/dalite-ng/venv/lib/python2.7/site-packages/django/views/generic/base.py" in dispatch
  89.         return handler(request, *args, **kwargs)
File "/home/jboc/programs/dalite-ng/venv/lib/python2.7/site-packages/django/views/generic/base.py" in get
  158.         context = self.get_context_data(**kwargs)
File "/home/jboc/programs/dalite-ng/peerinst/admin_views.py" in get_context_data
  239.         sums, rationale_data = get_question_rationale_aggregates(assignment, question)
File "/home/jboc/programs/dalite-ng/peerinst/admin_views.py" in get_question_rationale_aggregates
  154.                     'rationale': rationales[answer.chosen_rationale_id],

Exception Type: KeyError at /admin/peerinst/assignment_results/test=report/rationales/7
Exception Value: 16L
```

jbzdak commented on peerinst/admin_views.py:
```diff
@@ -86,6 +86,166 @@ def get_assignment_aggregates(assignment):
+
+    # Show rationales that changed answers from wrong to right, or right to wrong
+    answerchoice_correct = question.answerchoice_set.values_list('correct', flat=True)
+    correct_choices = list(itertools.compress(itertools.count(1), answerchoice_correct))
```
I have very hard time understanding this code, please explain what does it do.

My current understanding: 

1. answers is a list of answers
2. answerchoice_correct is a list with True if answer is correct and false otherwise for every `AnswerChoice` attached to question. 
3. correct_choices is a list of indices of answers where question is correct. 

Based on my unserstanding I have following doubt: `correct_choices` correspond to indices in `question.answerchoice_set` not to e.g. primary keys of `AnswerChoice` model, which means that using them on query `answers.exclude(first_answer_choice__in=correct_choices` won't work properly.

pomegranited replied:
`Answer.first_answer_choice` and `.second_answer_choice` aren't primary keys of `AnswerChoice`, they're indexes into the `Question.answerchoice_set` list. See [QuestionStartView.form_valid()](https://github.com/open-craft/dalite-ng/blob/jill/top-rationales/peerinst/views.py#L196) and [Question.is_correct()](https://github.com/open-craft/dalite-ng/blob/jill/top-rationales/peerinst/models.py#L184).

I have no idea why it was done that way!

jbzdak commented on peerinst/admin_views.py:
```diff
@@ -86,6 +86,166 @@ def get_assignment_aggregates(assignment):
+                group['rationales'][answer.chosen_rationale_id] = {
+                    'rationale': rationales[answer.chosen_rationale_id],
+                    'count': 0,
+                    'answers': [],
```
The `answers` entry in the report seems unused anywhere outside of this loop?

pomegranited replied:
Removed by [1191cea](https://github.com/open-craft/dalite-ng/pull/43/commits/1191cea2af37032acdab2227eed74100b78e9b5d)

jbzdak commented on peerinst/admin_views.py:
```diff
@@ -86,6 +86,166 @@ def get_assignment_aggregates(assignment):
+def get_question_rationale_aggregates(assignment, question):
+    """Get aggregate upvoted rationales for the given assignment and question.
+
+    This function returns a pair (sums, rationales), where 'sums' is a collections.Counter object
```
Result of this function is very complex object (pair of maps). Could we return namedtouple, so we could just take: 
```
report = get_question_rationale_aggregates(..)
report.total_upvoted
report.right_to_wrong
```


Then we could document each property separately

jbzdak commented on peerinst/admin_views.py:
```diff
@@ -86,6 +86,166 @@ def get_assignment_aggregates(assignment):
+    # Group by chosen_rational
+    rationales = {a.id: a for a in answers.all()}
+    rationales[None] = None
+    answer_groups = {
```
Wouldn't it be easier to just have three variables: `wrong_to_right`, `right_to_wrong`, `upvoted` instead od dictionary storing three keys?

pomegranited replied:
Improved by refactoring: [1191cea](https://github.com/open-craft/dalite-ng/pull/43/commits/1191cea2af37032acdab2227eed74100b78e9b5d)

jbzdak commented on peerinst/admin_views.py:
```diff
@@ -86,6 +86,166 @@ def get_assignment_aggregates(assignment):
+        for item in rationale_data:
+            count = item.get('count', 0)
+            rationale = item.get('rationale', 'None')
+            if isinstance(rationale, models.Answer):
```
Why rationales are mixed with Answers?

pomegranited replied:
Clarified with [1191cea R190](https://github.com/open-craft/dalite-ng/pull/43/commits/1191cea2af37032acdab2227eed74100b78e9b5d#diff-6fdfe2e57a137ae571d8a456bc39f4bdR190) - as noted below, rationales are Answers, but I really only needed to know if `rationale == None`.

jbzdak commented:
@pomegranited Overall could you please simplify code generating the report? Like add some helper functions, extract common logic and so forth.

pomegranited commented:
@jbzdak Thank you for your suggestions and for finding that 500.  I've refactored the code and hopefully made it clearer.  The refactor also removed the code that caused the 500, but could you please re-test?

Here's the background information that I learned while doing this task.  Sorry I didn't provide it earlier!

When a student answers a question, she:
* Creates an Answer instance which contains her first selected answer, and a rationale string.  So each Answer instance is actually a "Rationale" instance as well.
* The student is given a second try at answering the question, and is presented with a list of rationales given by other students (selected according to some algorithm) which support the student's first selected answer, and another answer.
* The student can choose to keep her current answer, and keep her rationale or choose another student's rationale.
* Or the student can choose a different answer using another student's rationale.
* If the student chooses another student's rationale, her Answer instance is updated to set `chosen_rationale` to that student's Answer instance.

The question can be configured to allow `sequential_review`, which also lets students upvote or downvote a list of rationales.

This task had three parts:

1. Show the most upvoted rationales, sorted by upvotes.
1. Show the chosen rationales which influenced answers that switched from wrong to right, sorted by number of times chosen.
1. Show the same as #2, but for the rationales which influenced answers that switched from right to wrong.

jbzdak commented on peerinst/admin_views.py:
```diff
@@ -86,6 +86,160 @@ def get_assignment_aggregates(assignment):
+        rows = []
+        for item in rationale_data:
+            count = item.get('count', 0)
+            rationale = item.get('rationale', 'None')
```
Should be `None` not `'None'` (without quotes), otherwise if below won't work.

pomegranited replied:
Fixed by [726fe02](https://github.com/open-craft/dalite-ng/pull/43/commits/726fe02b29124c23ad6b55daed9bc5c54c713104)

jbzdak commented on peerinst/admin_views.py:
```diff
@@ -86,6 +86,160 @@ def get_assignment_aggregates(assignment):
+
+    # Collect answers changed from wrong to right, and right to wrong
+    answerchoice_correct = question.answerchoice_set.values_list('correct', flat=True)
+    correct_choices = list(itertools.compress(itertools.count(1), answerchoice_correct))
```
Add some comment what correct_choices contain. Indexes of correct answers for the question

pomegranited replied:
Fixed by [6751b5d](https://github.com/open-craft/dalite-ng/pull/43/commits/6751b5d61da591eca8573ae8e97cb18283744441)

jbzdak commented on peerinst/admin_views.py:
```diff
@@ -86,6 +86,160 @@ def get_assignment_aggregates(assignment):
+                      .select_related('chosen_rationale'))
+
+    # Group answers by chosen_rationale
+    for key, answers in (('wrong_to_right', wrong_to_right),
```
I'd extract contents of this block to a helper function.

pomegranited replied:
Done with [2a84f48](https://github.com/open-craft/dalite-ng/pull/43/commits/2a84f483011c325b393a4d0a4ab46ad9c13b05bb#diff-6fdfe2e57a137ae571d8a456bc39f4bdR128)

jbzdak commented on peerinst/tests/test_admin_views.py:
```diff
@@ -60,3 +64,176 @@ def test_get_assignment_aggregates(self):
+                'upvoted': [
+                    {
+                        'count': 10,
+                        'rationale': models.Answer.objects.get(id=855),
```
This fails tests for me. This is a gotcha moment for Django. Anyways test fixtures are loaded only inside setUp method, and this line is executed beforehand --- during module loading. It works for you, because it queries your local database (test database is injected into settings after). I didn't load the fixture, and I got;

```
ImportError: Failed to import test module: peerinst.tests.test_admin_views
Traceback (most recent call last):
  File "/opt/python/2.7.11/lib/python2.7/unittest/loader.py", line 254, in _find_tests
    module = self._get_module_from_name(name)
  File "/opt/python/2.7.11/lib/python2.7/unittest/loader.py", line 232, in _get_module_from_name
    __import__(name)
  File "/home/jboc/programs/dalite-ng/peerinst/tests/test_admin_views.py", line 13, in <module>
    class AggregatesTestCase(TestCase):
  File "/home/jboc/programs/dalite-ng/peerinst/tests/test_admin_views.py", line 78, in AggregatesTestCase
    'rationale': models.Answer.objects.get(id=855),
  File "/home/jboc/programs/dalite-ng/venv/lib/python2.7/site-packages/django/db/models/manager.py", line 127, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/home/jboc/programs/dalite-ng/venv/lib/python2.7/site-packages/django/db/models/query.py", line 334, in get
    self.model._meta.object_name
DoesNotExist: Answer matching query does not exist.
```

pomegranited replied:
Gah, thanks for finding that!  Fixed with [ea0c2d8](https://github.com/open-craft/dalite-ng/pull/43/commits/ea0c2d894ea102a20b3cbe72989e48dfbe24ec94)

jbzdak commented on peerinst/tests/test_admin_views.py:
```diff
@@ -60,3 +64,176 @@ def test_get_assignment_aggregates(self):
+        )
+    )
+    def test_get_question_rationale_aggregates(self, assign_id, question_id, expected_sums, expected_rationales):
+        assignment = models.Assignment.objects.get(identifier=assign_id)
```
This kind of tests introduces some technical debt --- as it relies on complicated fixture, which is queried and then we have a result. If the fixture is altered in some way (due to another change or new migration) this test may fail. Also I have hard time understanding why the responses for assignment 30 are such and such. This is a known anti-pattern named [Mystery Guest](https://robots.thoughtbot.com/mystery-guest). 

However since most of tests in this file follow this pattern, and this is a time-boxed task we may just let it be, and hope our future-selves won't regret this decision.

pomegranited replied:
I hear you.. I was relieved that my data changes for these tests didn't break the existing tests.  But given the many relationships among the data required by this rationale report, I don't see any other way of testing it without elaborate fixture data.

jbzdak replied:
The usual way to fix it would be to use factory boy or something similar to create create only necessary initial data. However introducing factories to testing is definitely out of scope :)

jbzdak commented:
@pomegranited Thanks for the changes! I confirm that the 500 issue is no longer present. 

I left some more comments, and when you address them I'm mostly OK with this code then I'd like to re-test it manually.

pomegranited commented:
@jbzdak Thank you for your detailed review!  I'm much happier with the code now, hope you are too.

jbzdak commented:
I'm OK with the code, now some testing. 

Here is what I did: 

1. I created two test questions
2. And filled them, choosing "stick to my own rationale" and here is what the report shown: 

![selection_001](https://cloud.githubusercontent.com/assets/112872/14878156/fc6af064-0d21-11e6-8b3d-ddad39b25369.png)

I'm confused by the `(None)`, maybe just add: `Student stuck to his rationale`.

jbzdak commented:
@pomegranited :+1: After fixing the `(None)` issue.

pomegranited commented:
@jbzdak I've fixed that label, thanks!  


Do you know if I need more than one reviewer before merging this?
CC @bradenmacdonald

jbzdak commented:
@pomegranited AFAIK no this is internal repository owned by our team, so we only need one reviewer from the team.

bradenmacdonald commented:
@pomegranited This looks nice! Thanks for getting this done. One question that I couldn't tell from the code or the screenshots: Is there a limit to how many "top rationales" will be displayed? If there are 1,000 rationales, we probably don't want to display the "top 1,000 rationales"...

pomegranited commented:
@bradenmacdonald Addressed by PR #44.
